Fix PanZoomInteraction constraints

hellochar opened the pull request:
|| should've been an &&

hellochar committed:
hit "constrain both ends" codepath correctly

|| should've been an &&

hellochar commented:
Hmm this is breaking on x axis panzoom...

hellochar committed:
refactor panzoomconstraints math into its own file

add axis panzoom to pan_zoom_reverse.js

rename the basic "zoomAt" to "zoomOut"

hellochar committed:
fix names and add comments

hellochar committed:
fix lint; move constrainedTranslation into constraints file

hellochar committed:
fix lint

hellochar committed:
write better comment for constrainZoomValues

themadcreator commented on quicktests/overlaying/tests/interactions/pan_zoom_reverse.js:
```diff
@@ -29,6 +29,14 @@ function run(svg, data, Plottable) {
+
+    new Plottable.Interactions.PanZoom(null, yScale)
+        .attachTo(yAxis)
+        .setMinMaxDomainValuesTo(yScale);
```
Trying out this preview, i see this behavior:

![constrained-zoom-2](https://user-images.githubusercontent.com/569364/31693434-69e385c4-b353-11e7-972a-1d597a94d4d0.gif)

If you zoom an axes, you can use the plot to zoom all the way out. It imposes a constraint that doesn't match the scale extents.

themadcreator replied:
The behavior that would be most desirable, IMO, is to have the x axis zoom out while the y axis stays fixed. That is, constraining one axis doesn't necessarily constrain the other axis. Also what about multiple independent axes?

hellochar replied:
Yeah we had discussed this - I decided against it because users can already implement that behavior by creating separate PanZoomInteractions and attaching all of them to the same plot. If we make that the behavior of PZI, then we no longer support the users who want the current "constrain all scales" behavior.

themadcreator replied:
That may be possible for plottable users, but not for our react component. If we are going to use this axis zoom ourselves, won't our users have the same problem with no way to fix it?

hellochar replied:
We can do the same with our React Component, I was experimenting with it. We just rebuild an array of PZIs instead of a single one when the props change

themadcreator commented on src/interactions/panZoomConstraints.ts:
```diff
@@ -0,0 +1,176 @@
+    return { centerPoint, zoomAmount };
+  }
+
+  minDomainValue = minDomainValue == null ? -Infinity : minDomainValue;
```
won't `-Infinity` be wrong if the domain is reversed? maybe verify this by removing the max constraint from the pan_zoom_reverse preview

themadcreator replied:
otherwise, i like this approach to clean up this function

themadcreator reviewed (commented):
Questions

hellochar committed:
the .zoom() method now respects constraints

technically a breaking change (although the original behavior was never
explicitly defined).

hellochar committed:
account for domain reversal in minmax constraints

themadcreator commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -140,25 +133,44 @@ export class PanZoom extends Interaction {
    * By default, `centerValue` is the center of the x and y range of each scale.
    */
   public zoom(zoomAmount: number, centerValue?: Point) {
```
nit: given this is a functional change, it might be wise to hedge against upsetting users with an optional third arg to disable the constraint.

themadcreator reviewed (approved):
not gonna block but i still think the locked axis constraint is gonna be an issue later

hellochar committed:
add option to disable constraining to zoom method
